### PR TITLE
feat(*)!: 1.0 compatible

### PR DESCRIPTION
**Pull request by brooksmtownsend**
## Feature or Problem
This PR is updating the provider SDK to match 1.0-compatible topics, structures, etc.

Notably this does not yet include the wRPC implementation which is separately implemented and will be imported here as a module.

## Related Issues
Follow up https://github.com/wasmCloud/provider-sdk-go/issues/11

### Manual Verification
The health and shutdown topics were manually verified

### CODE REVIEW

**Review by thomastaylor312 — COMMENTED** [review]
Overall looking pretty good

File: host.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	EnableMetrics         *bool  `json:"enable_metrics,omitempty"`
+	EnableLogs            *bool  `json:"enable_logs,omitempty"`
+	ObservabilityEndpoint string `json:"observability_endpoint,omitempty"`
+	TracesEndpoint        string `json:"traces_endpoint,omitempty"`
+	MetricsEndpoint       string `json:"metrics_endpoint,omitempty"`
+	LogsEndpoint          string `json:"logs_endpoint,omitempty"`
+}
+
+type HostData struct {
+	HostID              string                    `json:"host_id,omitempty"`
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply host.go:15]: Yeah I think I copied all the omitempty values from the way we handle the struct in Rust with `serde(default)` macros, but they certainly should be there

File: examples/keyvalue-inmemory/main.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	"os/signal"
+	"syscall"
+
+	"github.com/wasmCloud/provider-sdk-go"
+	server "github.com/wasmCloud/provider-sdk-go/examples/keyvalue-inmemory/bindings"
+	"golang.org/x/sync/errgroup"
+)
+
+func main() {
+	providerGroup := new(errgroup.Group)
```
Comment (rvolosatovs) [inline examples/keyvalue-inmemory/main.go:17]: nit;
```suggestion
	var providerGroup errgroup.Group
```

File: examples/keyvalue-inmemory/main.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	p, err := provider.New(
+		provider.SourceLinkPut(handleNewSourceLink),
+		provider.TargetLinkPut(handleNewTargetLink),
+		provider.SourceLinkDel(handleDelSourceLink),
+		provider.TargetLinkDel(handleDelTargetLink),
+		provider.HealthCheck(handleHealthCheck),
+		provider.Shutdown(handleShutdown),
+	)
+	if err != nil {
+		panic(err)
```
Comment (rvolosatovs) [inline examples/keyvalue-inmemory/main.go:27]: nit; maybe replace this by a `log.Fatal`?
how about we put this logic into a `run() error`, such that `main` just calls `run` and handles the error gracefully? see https://github.com/wrpc/wrpc/blob/a779107b744f9f986739fd0046c08b546f167164/examples/go/keyvalue-server/cmd/keyvalue-mem-nats/main.go#L173-L177

File: examples/keyvalue-inmemory/main.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	"syscall"
+
+	"github.com/wasmCloud/provider-sdk-go"
+	server "github.com/wasmCloud/provider-sdk-go/examples/keyvalue-inmemory/bindings"
+	"golang.org/x/sync/errgroup"
+)
+
+func main() {
+	providerGroup := new(errgroup.Group)
+	p, err := provider.New(
```
Comment (rvolosatovs) [inline examples/keyvalue-inmemory/main.go:18]: This is missing unsubscribe logic to unsubscribe from the topics on exit, right?

File: examples/keyvalue-inmemory/build.sh [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+# Build the Go binary and package it into a par file
+# This is designed to be run from the root of the project
+go build ./cmd/keyvalue-inmemory
```
Comment (rvolosatovs) [inline examples/keyvalue-inmemory/build.sh:5]: this is missing a `go generate` to generate the bindings

File: host.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply host.go]: I think this should be fine

File: host.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+type HostData struct {
+	HostID             string `json:"host_id,omitempty"`
+	LatticeRPCPrefix   string `json:"lattice_rpc_prefix,omitempty"`
+	LatticeRPCUserJWT  string `json:"lattice_rpc_user_jwt,omitempty"`
+	LatticeRPCUserSeed string `json:"lattice_rpc_user_seed,omitempty"`
+	LatticeRPCURL      string `json:"lattice_rpc_url,omitempty"`
+	ProviderKey        string `json:"provider_key,omitempty"`
+	// NOTE: Rust is trying to pass a Vec<(String, String)> here, can we support that?
```
Comment (thomastaylor312) [inline host.go:21]: We should probably do a custom `UnmarshalJSON` function to handle this for now

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
All good to go except for parsing the proper type out of incoming host data

File: host.go (hunk 6 above)
Comment (brooksmtownsend) [reply host.go]: I did test this actually, and it works just fine 😄 Even though right now in the host we always pass a `vec![]`. In rust, passing `vec![("foo".to_string(), "bar".to_string())]` results in:
```
ProviderKey:pgz EnvValues:map[foo:bar] InstanceID:c31b57c6-a564-4c28-872e-4255cf5384cf
```

Comment (thomastaylor312) [reply host.go]: That is great!

**Review by thomastaylor312 — APPROVED** [review]

**Pull request merged**